stellar-contract-env-host/src/host.rs

```diff
@@ -1089,17 +1089,51 @@ impl CheckedEnv for Host {
 
     fn map_prev_key(&self, m: Object, k: RawVal) -> Result<RawVal, HostError> {
         let k = self.associate_raw_val(k);
-        self.visit_obj(m, |hm: &HostMap| match hm.get_prev(&k) {
-            Some((pk, pv)) => Ok(pk.to_raw()),
-            None => Ok(UNKNOWN_ERROR.to_raw()), //FIXME: replace with the actual status code
+        self.visit_obj(m, |hm: &HostMap| {
+            if let Some((pk, pv)) = hm.get_prev(&k) {
+                if *pk != k {
+                    Ok(pk.to_raw())
+                } else {
+                    if let Some((pk2, pv2)) = hm
+                        .clone()
+                        .extract(pk) // removes (pk, pv) and returns an Option<(pv, updated_map)>
+                        .ok_or_else(|| HostError::General("key not exist"))?
+                        .1
+                        .get_prev(pk)
+                    {
+                        Ok(pk2.to_raw())
+                    } else {
+                        Ok(UNKNOWN_ERROR.to_raw()) //FIXME: replace with the actual status code
+                    }
+                }
+            } else {
+                Ok(UNKNOWN_ERROR.to_raw()) //FIXME: replace with the actual status code
+            }
         })
     }
 
     fn map_next_key(&self, m: Object, k: RawVal) -> Result<RawVal, HostError> {
         let k = self.associate_raw_val(k);
-        self.visit_obj(m, |hm: &HostMap| match hm.get_next(&k) {
-            Some((pk, pv)) => Ok(pk.to_raw()),
-            None => Ok(UNKNOWN_ERROR.to_raw()), //FIXME: replace with the actual status code
+        self.visit_obj(m, |hm: &HostMap| {
+            if let Some((pk, pv)) = hm.get_next(&k) {
+                if *pk != k {
+                    Ok(pk.to_raw())
+                } else {
+                    if let Some((pk2, pv2)) = hm
+                        .clone()
+                        .extract(pk) // removes (pk, pv) and returns an Option<(pv, updated_map)>
```

Comment:
Oh interesting, so we're removing the key from the OrdMap, so that we can find the next key? That will result in a copy of the OrdMap is that right? 🤔 The main reason we're fixing these functions, besides the fact their broken, is so we can use them in an SDK to avoid making a copy of the map, because at the moment the SDK just copies the host object and removes the key itself. So it feels like we're shifting the existing fix upstream, which is better because maybe we can find a better fix someday host side, but is there not another way we can get the next without a copy?

Reply:
Unfortunately I couldn't find a different way to do it without cloning the OrdMap. One way I was trying to do it was to get the iterator on the input key in the map then call `next` (or `next_back`) on it. However, there are very few methods that return iterator to map entry (which seems very different from C++ containers and iterators). 
Anyways, I believe the performance should not be a big concern. The im-rc does "lazy cloning" which only happens when the data is modified, and our worse case here is deleting one entry, which means cloning is very limited (that node plus the parent nodes). The big difference from doing it on the SDK side is no new host object will be created. The overall performance should still on the order of `log(n)` and should be very similar to (if not slightly better than) a `map_del` (since no new host object is created).
I'm planning to wire up metering on the host functions soon, so hopefully we will have a more concrete answer to the performance soon. 
(I've also added a much longer comment in the code explaining some of the rationale)

Reply:
Makes sense. Also, even if there was no immediate to doing this in the host, if we ever find a better way to solve this in the future at least having it in the host means we can change it seamlessly. :tada: